instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/ApiGatewayProxyRequest.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.awslambda.v1_0;
+
+import com.amazonaws.serverless.proxy.model.Headers;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.module.afterburner.AfterburnerModule;
+import java.io.InputStream;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ApiGatewayProxyRequest {
+
+  private static final Logger log = LoggerFactory.getLogger(ApiGatewayProxyRequest.class);
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  static {
+    OBJECT_MAPPER.registerModule(new AfterburnerModule());
+  }
+
+  private final Headers headers;
+
+  private ApiGatewayProxyRequest(Headers headers) {
```

Comment:
This is unfortunate but seems to be the only way to support streams, oh well.

For non-streams, will we use the event types?
https://github.com/aws/aws-lambda-java-libs/blob/master/aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/APIGatewayProxyRequestEvent.java

Reply:
Though I wonder if supporting stream is actually important. Looking at this, it seems like the body is encoded into JSON anyways so there is no benefit in the stream handler - so maybe it's a reasonable and practical limitation for us to just detect that event type?

Reply:
Yes, events for non-streams should be doable, at least at the wrapper level. Handler gets argument type from a child class so in this case afair it's not possible to get the headers if arg is not the event type (correct me if I'm wrong). 

With regards to streams, now we have safeguards (only if propagators configured) and current implementation of the AWS proxy uses ButeArrayInputStream which is markable, so I'd say it's safe to leave it (as a drop-in method for customers already suing streams handler). 
If you agree, we can merge this PR and I'll file another one for non-streams wrapper.